### PR TITLE
Add an example for a dhcp server using charrua-core.

### PR DESCRIPTION
**Pull request by haesbaert**
This adds the basic code needed for making a dhcp server as discussed in:

http://lists.xenproject.org/archives/html/mirageos-devel/2015-10/msg00039.html

### CODE REVIEW

**Comment by yomimono** [comment]
I'm surprised that `unikernel.ml` does its own parsing from the Ethernet layer on up.  Is it not possible to use the Stack module to provide access at a higher layer, so that all of this code isn't duplicated in the unikernel?

**Comment by haesbaert** [comment]
If I use the Stack module I end up getting the payload of udp, while in
fact I need the whole packet. Or did I miss something ?

**Comment by avsm** [comment]
Yeah there is no raw frame access from Stack (this needs plumbing raw sockets for the Socket backend, hence doesn't exist yet)

**Comment by mor1** [comment]
Hi @haesbaert -- wondering how adding support for 4.01 was going? That looks like the only thing blocking this to me... :)

**Comment by haesbaert** [comment]
Sorry, been super busy, I plan to look at this friday or saturday.

**Comment by mor1** [comment]
No problem! Just checking it wasn't blocked on me (for a change :)

On 28 October 2015 at 17:29, Christiano Haesbaert <notifications@github.com>
wrote:

> Sorry, been super busy, I plan to look at this friday or saturday.
>
> On Wednesday, 28 October 2015, Richard Mortier <notifications@github.com>
> wrote:
>
> > Hi @haesbaert <https://github.com/haesbaert> -- wondering how adding
> > support for 4.01 was going? That looks like the only thing blocking this
> to
> > me... :)
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/mirage/mirage-skeleton/pull/105#issuecomment-151906511>
> > .
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage-skeleton/pull/105#issuecomment-151919021>
> .
>



-- 
Richard Mortier
mort@cantab.net

**Comment by haesbaert** [comment]
I've updated the diff, the build will fail until the opam packages are integrated in
https://github.com/ocaml/opam-repository/pull/5115

**Comment by haesbaert** [comment]
Should be fine now.

File: dhcp/config.ml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+open Mirage
+
+let main = foreign "Unikernel.Main" (console @-> kv_ro @-> network @-> clock @-> job)
+
+let disk = crunch "files"
+
+(* TODO: workaround a bug in the command-line tool by adding Clock
+   for Unix (this is pulled in as an implicit dependency in Xen) *)
```
Comment (hannesm) [inline dhcp/config.ml]: is this reported somewhere, that `mirage-clock-unix` needs to be manually pulled in? does it happen with mirage-master or the latest release? @Drup @samoht

File: dhcp/unikernel.ml [hunk 2]
```diff
@@ -0,0 +1,85 @@
+open V1_LWT
+open Lwt
+
+(* IP Configuration, all you need besides dhcpd.conf. *)
+let ipaddr = Ipaddr.V4.of_string_exn "192.168.1.5"
```
Comment (hannesm) [inline dhcp/unikernel.ml]: shouldn't this be passed in from `config.ml`?

File: dhcp/config.ml (hunk 1 above)
Comment (Drup) [inline dhcp/config.ml]: This is ... ancient. I think @haesbaert copied the config.ml from stack, where this was present. It's not valid anymore.

File: dhcp/unikernel.ml (hunk 2 above)
Comment (haesbaert) [inline dhcp/unikernel.ml]: On Wednesday, 11 November 2015, Hannes Mehnert <notifications@github.com>
wrote:

> In dhcp/unikernel.ml
> <https://github.com/mirage/mirage-skeleton/pull/105#discussion_r44581922>:
>
> > @@ -0,0 +1,85 @@
> > +open V1_LWT
> > +open Lwt
> > +
> > +(* IP Configuration, all you need besides dhcpd.conf. *)
> > +let ipaddr = Ipaddr.V4.of_string_exn "192.168.1.5"
>
> shouldn't this be passed in from config.ml?
>
> —
>
>
>

Yep, I couldn't figure how, should I pass a module with just that to the
main in unikernel ?

File: dhcp/config.ml (hunk 1 above)
Comment (haesbaert) [inline dhcp/config.ml]: On Thursday, 12 November 2015, Gabriel Radanne <notifications@github.com>
wrote:

> In dhcp/config.ml
> <https://github.com/mirage/mirage-skeleton/pull/105#discussion_r44648033>:
>
> > @@ -0,0 +1,21 @@
> > +open Mirage
> > +
> > +let main = foreign "Unikernel.Main" (console @-> kv_ro @-> network @-> clock @-> job)
> > +
> > +let disk = crunch "files"
> > +
> > +(* TODO: workaround a bug in the command-line tool by adding Clock
> > +   for Unix (this is pulled in as an implicit dependency in Xen) *)
>
> This is ... ancient. I think @haesbaert <https://github.com/haesbaert>
> copied the config.ml from stack, where this was present. It's not valid
> anymore.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage-skeleton/pull/105/files#r44648033>.
>
Correct, I copied over from the ethif skeleton.
I will clean it up.

File: dhcp/unikernel.ml (hunk 2 above)
Comment (hannesm) [inline dhcp/unikernel.ml]: I'm not entirely sure, maybe someone of @samoht @Drup @avsm has an answer?

Comment (samoht) [inline dhcp/unikernel.ml]: it'd be very easy to do with the new functoria, see https://github.com/Drup/mirage-skeleton/blob/functoria/hello/config.ml#L3-L5

Comment (amirmc) [inline dhcp/unikernel.ml]: Does that mean we should wait until functoria is properly released before adding this?

Comment (haesbaert) [inline dhcp/unikernel.ml]: If there is no other way, I would vote for pushing as it is and then fixing
once functoria is released.

Btw: im out of town until monday and wont have access to my machines.

Comment (mor1) [inline dhcp/unikernel.ml]: Agree-- I'd not hold this up for functoria to merge.

If there are things in (eg) ethif example that needed updating for this example, then possible skeleton needs some general TLC again to bring it back up to date.

(Indeed, having just pushed DNS support into mirage/mirage-seal, I'm tempted to suggest we might think about merging these two repos, as that might make it more likely that code will stay up-to-date with latest Mirage library support... But that might not be a good idea, and needs some thought :)

**Comment by mor1** [comment]
Ok-- as I don't think it's worth holding this up for functoria, is there anything else blocking accpeting this now?
@haesbaert @hannesm @samoht 

@haesbaert  Any chance we could get a blog post describing this please? Doesn't have to be particularly long! :)  (I think @amirmc was looking for one, but either please tell me to shut up if you've already provided one!)

**Comment by amirmc** [comment]
Yup, I'm still after a blog post :)

**Comment by hannesm** [comment]
I think this is fine to merge.  the `master` branch (of mirage-skeleton) is supposed to work with `mirage` from `opam-repository`; whereas `mirage-dev` branch here is for mirage from master... thus, no need to wait for functoria here (but once functoria is released, certainly this should be updates :)  also, travis is happy!

**Comment by amirmc** [comment]
This PR looks like it’s going into `master` branch — which means it needs to work with the mirage that’s in opam-repo (I assume that’s already the case).

If @haesbaert is happy to work on a functoria version in parallel, then a PR to the `mirage-dev` branch would be fine (and could be merged into `master` branch when functoria is released). That wouldn’t affect this PR being merged and would actually be useful to anyone else who’s looking for examples using the new lib.

File: dhcp/unikernel.ml (hunk 2 above)
Comment (Drup) [inline dhcp/unikernel.ml]: > If there are things in (eg) ethif example that needed updating for this example, then possible skeleton needs some general TLC again to bring it back up to date.

I did that when in my functoria PR.

Also, now, mirage's (and functoria's) travis run mirage-skeleton.

**Comment by haesbaert** [comment]
ok I've updated the diff to remove the old xen workaround, if I understood this should go into master and then I can work on functoria in mirage-dev or wait until the release to convert master.

**Comment by amirmc** [comment]
I’d be happy to see a functoria version in mirage-dev (as well as as some user feedback on functoria) :)

@mor1 are you happy to merge this?  @samoht, any comments?

**Comment by haesbaert** [comment]
Great, I can do that, I want to work on the blogpost first, sorry I've been out lately, had holidays and then other stuff.

**Comment by amirmc** [comment]
Fabulous! Yes, I think a blog post to go along with the addition to mirage-skeleton should be next.

Thanks for your time on this :)

**Comment by hannesm** [comment]
how comes that 30de04b is a part of this PR?

**Comment by haesbaert** [comment]
oh I had missed that, I've no idea how that snuck in, leme check.

**Comment by haesbaert** [comment]
Zefixed, missed a rebase.

**Comment by mor1** [comment]
Merged. Thanks!

**Comment by haesbaert** [comment]
\o/

**Comment by amirmc** [comment]
:100:

**Pull request merged**